# The size and shape of a CSV file

<div class="alert alert-info"><img src="../images/hhicon.png" width="50px" style="vertical-align: bottom; margin-right: 10px;">Remember to run each cell using <b>Shift+Enter</b>!</div>

## Opening a CSV file

In [66]:
import pandas as pd
import altair as alt

Using Pandas we can open a CSV file from a url – no need to download it first. To begin with, we open the CSV file and give it the name `df` (for 'DataFrame').

In this case, we're opening the [Brisbane Hospital registers of deaths 1899 to 1913](https://data.qld.gov.au/dataset/34ff0384-e09a-4ba0-8ea4-f54534e1aace) from the Queensland State Archives.

In [43]:
df = pd.read_csv('https://www.data.qld.gov.au/dataset/34ff0384-e09a-4ba0-8ea4-f54534e1aace/resource/265df65a-372e-4f90-a6b8-5a47c30fc740/download/brisbane-hospital-deaths-1899-1913.csv')

From now on we can just use `df` when we want to refer to the dataset. 

Pandas dataframes have lots of useful built-in tricks for exploring and analysing the dataset. We can use `.head()` to show us the first few rows.

In [44]:
df.head()

Last name      Given names   Age    No. Date of death Prev sys  Item ID  \
0    ABBOTT           Albert  47.0  146.0    1907-06-24  HOS1/18     2887   
1      ABEL     Adam Phillip  45.0  205.0    1913-08-04  HOS1/18     2887   
2      ABEL  Johanna Pauline  40.0  283.0    1907-11-19  HOS1/18     2887   
3   ABRAHAM           August  71.0  190.0    1912-07-09  HOS1/18     2887   
4   ABRAHAM           Kubeel  30.0  239.0    1906-11-27  HOS1/18     2887   

  Microfilm number                          Index name  \
0            Z1168  Brisbane Hospital deaths 1899-1913   
1            Z1168  Brisbane Hospital deaths 1899-1913   
2            Z1168  Brisbane Hospital deaths 1899-1913   
3            Z1168  Brisbane Hospital deaths 1899-1913   
4            Z1168  Brisbane Hospital deaths 1899-1913   

                                         Description  \
0  Registers kept by the Brisbane Hospital record...   
1  Registers kept by the Brisbane Hospital record...   
2  Registers kept by the Brisbane Hospital record...   
3  Registers kept by the Brisbane Hospital record...   
4  Registers kept by the Brisbane Hospital record...   

                                              Source  
0  http://www.archivessearch.qld.gov.au/Search/It...  
1  http://www.archivessearch.qld.gov.au/Search/It...  
2  http://www.archivessearch.qld.gov.au/Search/It...  
3  http://www.archivessearch.qld.gov.au/Search/It...  
4  http://www.archivessearch.qld.gov.au/Search/It...

<div class="alert alert-info"><img src="../images/hhicon.png" width="50px" style="vertical-align: bottom; margin-right: 10px;">Try putting a number in the brackets after <code>.head</code>. What happens? (Remember to hit <b>Shift+Enter</b> to run the cell once you've changed it!)</div>

## How big is the dataset?

We can use `.shape` to give us the number of rows and columns in the dataset.

In [45]:
df.shape

(4866, 11)

So this dataset has 4,866 rows.

## Get the values of a single column

To get the values of a single column (what Pandas calls a 'series'), just supply the column name in square brackets.

In [46]:
df['Last name']

0        ABBOTT
1          ABEL
2          ABEL
3       ABRAHAM
4       ABRAHAM
         ...   
4861    YOURELL
4862        YUE
4863    ZALLOTI
4864      ZERBE
4865    ZILLMAN
Name: Last name, Length: 4866, dtype: object

Note that for ease of display, Pandas shows you just the first and last values.

## Get the values of a single row

Rows in a dataframe are indexed – they're the numbers you can see in the left hand column. You can access any row by supplying it's index number to `.iloc`.

In [47]:
df.iloc[4]

Last name                                                     ABRAHAM
Given names                                                    Kubeel
Age                                                                30
No.                                                               239
Date of death                                              1906-11-27
Prev sys                                                      HOS1/18
Item ID                                                          2887
Microfilm number                                                Z1168
Index name                         Brisbane Hospital deaths 1899-1913
Description         Registers kept by the Brisbane Hospital record...
Source              http://www.archivessearch.qld.gov.au/Search/It...
Name: 4, dtype: object

<div class="alert alert-info"><img src="../images/hhicon.png" width="50px" style="vertical-align: bottom; margin-right: 10px;">Try changing the number in the square brackets! What happens?</div>

## Get all the rows that match a particular value

We can also supply conditions to `.iloc` to filter rows by a particular column and value. So to find all the rows where the `Last name` was 'ABRAHAM'...

In [48]:
df.loc[df['Last name'] == 'ABRAHAM']

Last name Given names   Age    No. Date of death Prev sys  Item ID  \
3   ABRAHAM      August  71.0  190.0    1912-07-09  HOS1/18     2887   
4   ABRAHAM      Kubeel  30.0  239.0    1906-11-27  HOS1/18     2887   

  Microfilm number                          Index name  \
3            Z1168  Brisbane Hospital deaths 1899-1913   
4            Z1168  Brisbane Hospital deaths 1899-1913   

                                         Description  \
3  Registers kept by the Brisbane Hospital record...   
4  Registers kept by the Brisbane Hospital record...   

                                              Source  
3  http://www.archivessearch.qld.gov.au/Search/It...  
4  http://www.archivessearch.qld.gov.au/Search/It...

<div class="alert alert-info"><img src="../images/hhicon.png" width="50px" style="vertical-align: bottom; margin-right: 10px;">Try changing the value of the name – what about 'SMITH'? Remember to keep the quotes around the name.</div>

## Find (or exclude) missing values

To find all rows where `Age` is not available, we can use `.iloc` with `.isnull`.

In [49]:
df.loc[df['Age'].isnull()]

Last name Given names  Age  No. Date of death Prev sys  Item ID  \
18         ADAMS       Annie  NaN  NaN    1900-10-11  HOS1/18     2887   
19         ADAMS        Jane  NaN  NaN    1902-07-05  HOS1/18     2887   
20         ADAMS         Reg  NaN  NaN    1913-04-03  HOS1/18     2887   
59         AHERN     Patrick  NaN  NaN    1899-09-07  HOS1/18     2887   
72    ALEXANDROF        Alex  NaN  NaN    1911-01-17  HOS1/18     2887   
...          ...         ...  ...  ...           ...      ...      ...   
4834      WRIGHT       Henry  NaN  NaN    1899-07-28  HOS1/18     2887   
4839       WYATT      Norman  NaN  NaN    1907-03-25  HOS1/18     2887   
4842       WYNNE        Mary  NaN  NaN    1908-09-13  HOS1/18     2887   
4844      YARROW       Alice  NaN  NaN    1902-12-09  HOS1/18     2887   
4860       YOUNG       Henry  NaN  NaN    1906-09-29  HOS1/18     2887   

     Microfilm number                          Index name  \
18              Z1168  Brisbane Hospital deaths 1899-1913   
19              Z1168  Brisbane Hospital deaths 1899-1913   
20              Z1168  Brisbane Hospital deaths 1899-1913   
59              Z1168  Brisbane Hospital deaths 1899-1913   
72              Z1168  Brisbane Hospital deaths 1899-1913   
...               ...                                 ...   
4834            Z1168  Brisbane Hospital deaths 1899-1913   
4839            Z1168  Brisbane Hospital deaths 1899-1913   
4842            Z1168  Brisbane Hospital deaths 1899-1913   
4844            Z1168  Brisbane Hospital deaths 1899-1913   
4860            Z1168  Brisbane Hospital deaths 1899-1913   

                                            Description  \
18    Registers kept by the Brisbane Hospital record...   
19    Registers kept by the Brisbane Hospital record...   
20    Registers kept by the Brisbane Hospital record...   
59    Registers kept by the Brisbane Hospital record...   
72    Registers kept by the Brisbane Hospital record...   
...                                                 ...   
4834  Registers kept by the Brisbane Hospital record...   
4839  Registers kept by the Brisbane Hospital record...   
4842  Registers kept by the Brisbane Hospital record...   
4844  Registers kept by the Brisbane Hospital record...   
4860  Registers kept by the Brisbane Hospital record...   

                                                 Source  
18    http://www.archivessearch.qld.gov.au/Search/It...  
19    http://www.archivessearch.qld.gov.au/Search/It...  
20    http://www.archivessearch.qld.gov.au/Search/It...  
59    http://www.archivessearch.qld.gov.au/Search/It...  
72    http://www.archivessearch.qld.gov.au/Search/It...  
...                                                 ...  
4834  http://www.archivessearch.qld.gov.au/Search/It...  
4839  http://www.archivessearch.qld.gov.au/Search/It...  
4842  http://www.archivessearch.qld.gov.au/Search/It...  
4844  http://www.archivessearch.qld.gov.au/Search/It...  
4860  http://www.archivessearch.qld.gov.au/Search/It...  

[334 rows x 11 columns]

So there are 334 rows that don't have a value for `Age`. Note that Pandas represents a missing value as `NaN`.

If we want to get all the available values for `Age` (without the `NaN`s), we can use `.dropna`.

In [50]:
df['Age'].dropna()

0       47.0
1       45.0
2       40.0
3       71.0
4       30.0
        ... 
4861    43.0
4862    63.0
4863    46.0
4864    71.0
4865    44.0
Name: Age, Length: 4532, dtype: float64

## Find the youngest and oldest ages

Using `.min()` and `.max()` we can find the smallest and largest values in a field.

In [51]:
df['Age'].min()

0.0

In [52]:
df['Age'].max()

96.0

To find the index of the row with the largest value we can use `.idxmax()`. Then we can give that value to `.iloc` to get the details of that row. So who was the 96-year-old patient?

In [53]:
df.loc[df['Age'].idxmax()]

Last name                                                     O KEEFE
Given names                                                      Mary
Age                                                                96
No.                                                                11
Date of death                                              1904-01-15
Prev sys                                                      HOS1/18
Item ID                                                          2887
Microfilm number                                                Z1168
Index name                         Brisbane Hospital deaths 1899-1913
Description         Registers kept by the Brisbane Hospital record...
Source              http://www.archivessearch.qld.gov.au/Search/It...
Name: 3338, dtype: object

We can even find out the mean age at death – we just use `.mean()`.

In [54]:
df['Age'].mean()

45.18556928508385

<div class="alert alert-info"><img src="../images/hhicon.png" width="50px" style="vertical-align: bottom; margin-right: 10px;">A challenge! Can you use what you've learnt on this page to find all records where the age at death was recorded as 0? The answer is hidden below – click on the dots to reveal it.</div>

In [ ]:
df.loc[df['Age'] == 0]

## What format are the dates in?

In [63]:
df['date'] = pd.to_datetime(df['Date of death'], errors='coerce')

In [65]:
df.loc[df['date'].isnull()]

Last name Given names  Age  No. Date of death Prev sys  Item ID  \
144     ASHTON       Daisy  NaN  NaN           NaN  HOS1/18     2887   
3380   OGILVIE      Hannah  NaN  NaN        Apr-08  HOS1/18     2887   

     Microfilm number                          Index name  \
144             Z1168  Brisbane Hospital deaths 1899-1913   
3380            Z1168  Brisbane Hospital deaths 1899-1913   

                                            Description  \
144   Registers kept by the Brisbane Hospital record...   
3380  Registers kept by the Brisbane Hospital record...   

                                                 Source date  
144   http://www.archivessearch.qld.gov.au/Search/It...  NaN  
3380  http://www.archivessearch.qld.gov.au/Search/It...  NaT

## What's next?

Using this notebook as a template, try exploring the size and shape of a different CSV file. There's plenty to choose from in this [GLAM data list](https://glam-workbench.github.io/glam-datasets-from-gov-portals/). Copy the url of the CSV file and paste it into the `.read_csv()` cell at the top. Of course, the column names are likely to be different, so you'll have to adjust them to suit.